lib/src/default_index/revset_engine.rs

```diff
@@ -95,13 +95,13 @@ impl<T: InternalRevset + ?Sized> InternalRevset for Box<T> {
     }
 }
 
-pub struct RevsetImpl<I> {
+pub struct RevsetImpl<'a> {
     inner: Box<dyn InternalRevset>,
-    index: I,
+    index: CompositeIndex<'a>,
```

Comment:
Well, I made this change so that a `Revset` computed for a `ReadonlyIndex` can be `'static`.

It's not needed right now, but will be useful to cache a `Template<Commit>` of the commit summary instance for example.

Reply:
Template<Commit> deals with a complete separate template language, right? How are these two things related?

I tried to preserve `RevsetImpl<>` storing the index but it became way too complicated due to the InternalRevset -> ExternalRevset nesting, and I imagine exposing `InternalRevset` as a public API for the extension to implement would be too drastic a change.  I'm open to suggestions but this seemed like the simplest solution.

Reply:
> Template deals with a completely separate template language, right? How are these two things related?

We'll probably add `revsets()` template function (or something like `immutable` that depends on revset), and its lifetime will be derived from the repo and index.

> it became way too complicated due to the InternalRevset -> ExternalRevset nesting, and I imagine exposing `InternalRevset` as a public API for the extension to implement would be too drastic a change.

Yeah, that's the problem. It's too general, I think. If the extension function is filter-like, the API for the default revset engine could be something like `Fn(&Commit) -> bool`. If it's aggregation-like, it might be `Fn(Box<dyn Iterator<Item = CommitId>>) -> Box<dyn Iterator<Item = CommitId>>`?

If evaluation strategies like these are defined by the core revset, we can also remove the `transform` hook point. I don't think `transform` can be reasonably made by both extension and core because it will need a global knowledge about the current tree state.